Custom view components

ofekcMoveo opened the pull request:
add a custom button to the app
add a custom header component

ofekcMoveo committed:
add search screen ui add search results iu + VC + viewModels
add functions to ArticleRepository which brings articles by keywords from the api

ofekcMoveo committed:
add pagination
connect actions to back buttons, clear button and remove buttons in the cells

ofekcMoveo committed:
changes in pagination logic

ofekcMoveo committed:
merge branch Alamofire to branch Search_Screen_and_Pagination

ofekcMoveo committed:
solve conflicts in project file

ofekcMoveo committed:
add colorPalleteNames struct
add reading api key from json file in the AppDelegate file
add .gitignore file
little changes in the controllers and view models

ofekcMoveo committed:
make homepage view model a singleton

ofekcMoveo committed:
add extension to String in order to send string as error description
change name of buttons in searchScreen and latestSearchCell
make ArticlesRepository a singleton

ofekcMoveo committed:
add images to assets
add model RecentSearches

ofekcMoveo committed:
add a search repository

ofekcMoveo committed:
delete an unnecessary decleration

ofekcMoveo committed:
custom a the buttons in the app

ofekcMoveo committed:
fix custom button

ofekcMoveo committed:
add custom header component

ofekcMoveo committed:
add TextFieldWithValidation component
add SignupOrLogin component
add logo header type

ofekcMoveo committed:
update gitignore file

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -8,26 +8,40 @@
         super.viewDidLoad()
         self.navigationController?.isNavigationBarHidden = true
         
+        headerView = HeaderView(frame: CGRect(x: 0, y: 0, width: homepageView.frame.width, height: 95), headerType: .mainHeader)
```
I'm curious why do you choose to add views programmatically in the code and not through the interface builder?\

Personally I prefer the opposite because then we know that when we look at the screen in the storyboard we can already know about all the elements that are in the page, and not some are in the storyboard and some are in the code. Consistency is important

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -8,26 +8,40 @@
+    private func setHeaderViewConstraints() {
+        headerView!.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            headerView!.leadingAnchor.constraint(equalTo: homepageView.leadingAnchor, constant: 0),
```
This relates to the comment above -  if you add the view inside the interface builder then you can already add all the constrains over there and have less code inside your VC. Also, personally I think it's very convenient to be able to see the constraints marked on the storyboard as it helps visualize them better in case they need to be changed

noamkuMoDev commented on Dispatcher/UI/Homepage/HomepageViewController.swift:
```diff
@@ -8,26 +8,40 @@
+    }
+    
     private func setupTableView() {
+        headerView?.delegate = self
```
The function is "setupTableView" so we want to do only tableView related stuff inside of it.
This line would be better placed inside the viewDidLoad

noamkuMoDev commented on Dispatcher/Views/ArticlesTableCells/ArticleCell.swift:
```diff
@@ -24,39 +24,62 @@ class ArticleCell: UITableViewCell {
+            iconPosition: .end
+        )
+
+        dispatchButton!.addTarget(self, action: #selector(navigateToDisptachPressed), for: .touchUpInside)
```
Usually we want to be informed about the press action on the button through a delegate and not by adding a touchUp selector to the button from the screen in which the button is implemented.

In case we have a more complex button and we want to perform an action on the button once we press it, it's easier to manage it inside the button itself and not from outside (from the VC that holds it)

ofekcMoveo committed:
add header view from story board
move the target function of the button to the button class and added a delegate
